### PR TITLE
feat!: add OCI artifact manifest type and handlers

### PR DESCRIPTION
**Pull request by wangxiaoxuan273**
Part 1/n of issue #21 

+ **Add OCI Artifact manifest type**: This was achieved by updating `github.com/opencontainers/image-spec` module to `v1.1.0-rc2`. 

+ **Add OCI Artifact manifest handlers**: I need to clean up the idea of where to put this.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -26,7 +26,7 @@ require (
 	github.com/mitchellh/osext v0.0.0-20151018003038-5e2d6d41470f // indirect
 	github.com/ncw/swift v1.0.47
 	github.com/opencontainers/go-digest v1.0.0
-	github.com/opencontainers/image-spec v1.0.2
+	github.com/opencontainers/image-spec v1.1.0-rc2.0.20221013174636-8159c8264e2e
```
Comment (shizhMSFT) [inline go.mod:29]: Can we pin to `v1.1.0-rc2` instead of a commit?

File: registry/handlers/manifests.go [hunk 2]
```diff
@@ -322,7 +333,7 @@ func (imh *manifestHandler) PutManifest(w http.ResponseWriter, r *http.Request)
 		return
 	}
 
-	isAnOCIManifest := mediaType == v1.MediaTypeImageManifest || mediaType == v1.MediaTypeImageIndex
+	isAnOCIManifest := mediaType == v1.MediaTypeImageManifest || mediaType == v1.MediaTypeArtifactManifest || mediaType == v1.MediaTypeImageIndex
```
Comment (shizhMSFT) [inline registry/handlers/manifests.go:336]: nit: we may refactor it using a switch case.

File: go.mod (hunk 1 above)
Comment (wangxiaoxuan273) [reply go.mod:22]: Done.

File: registry/handlers/manifests.go (hunk 2 above)
Comment (wangxiaoxuan273) [reply registry/handlers/manifests.go]: Done.

File: registry/handlers/manifests.go [hunk 3]
```diff
@@ -156,18 +160,25 @@ func (imh *manifestHandler) GetManifest(w http.ResponseWriter, r *http.Request)
 	manifestList, isManifestList := manifest.(*manifestlist.DeserializedManifestList)
 	if isSchema2 {
 		manifestType = manifestSchema2
-	} else if _, isOCImanifest := manifest.(*ocischema.DeserializedManifest); isOCImanifest {
-		manifestType = ociSchema
+	} else if ociManifest, isOCImanifest := manifest.(*ocischema.DeserializedManifest); isOCImanifest {
+		if ociManifest.MediaType == v1.MediaTypeImageManifest {
+			manifestType = ociImageSchema
+		} else if ociManifest.MediaType == v1.MediaTypeArtifactManifest {
```
[the remark this answers is not in the document]
Comment (wangxiaoxuan273) [reply registry/handlers/manifests.go:166]: Thanks for pointing out. This is crucial.

File: registry/handlers/manifests.go [hunk 4]
```diff
@@ -82,7 +83,7 @@ type manifestHandler struct {
 
 // GetManifest fetches the image manifest from the storage backend, if it exists.
```
Comment (qweeah) [inline registry/handlers/manifests.go:84]: nit: `image` also need to be removed from comments

File: registry/handlers/manifests.go [hunk 5]
```diff
@@ -156,18 +160,25 @@ func (imh *manifestHandler) GetManifest(w http.ResponseWriter, r *http.Request)
 	manifestList, isManifestList := manifest.(*manifestlist.DeserializedManifestList)
 	if isSchema2 {
 		manifestType = manifestSchema2
-	} else if _, isOCImanifest := manifest.(*ocischema.DeserializedManifest); isOCImanifest {
-		manifestType = ociSchema
+	} else if ociManifest, isOCImanifest := manifest.(*ocischema.DeserializedManifest); isOCImanifest {
```
Comment (Wwwsylvia) [inline registry/handlers/manifests.go:163]: nit: isOCI**M**anifest?

File: registry/handlers/manifests.go [hunk 6]
```diff
@@ -156,18 +160,25 @@ func (imh *manifestHandler) GetManifest(w http.ResponseWriter, r *http.Request)
...
+		if ociManifest.MediaType == v1.MediaTypeImageManifest {
+			manifestType = ociImageSchema
+		} else if ociManifest.MediaType == v1.MediaTypeArtifactManifest {
+			manifestType = ociArtifactSchema
+		}
```
Comment (Wwwsylvia) [inline registry/handlers/manifests.go:168]: nit: Do we want to refactor these using switch-case?

**Comment by wangxiaoxuan273** [comment]
After thinking I now decided that I should not put handlers for artifact in the `manifest.go` file (with the manifest handler). Instead I should have a separate handler for artifact in the `handlers` package, and a separate artifact package to provide the necessary support. Therefore I'll revoke my previous changes in the `manifest.go` file.

File: registry/handlers/manifests.go (hunk 4 above)
Comment (wangxiaoxuan273) [reply registry/handlers/manifests.go]: Removed.

File: registry/handlers/manifests.go (hunk 5 above)
Comment (wangxiaoxuan273) [reply registry/handlers/manifests.go]: Renamed `isOCIImageManifest.`

File: registry/handlers/manifests.go (hunk 6 above)
Comment (wangxiaoxuan273) [reply registry/handlers/manifests.go]: I'll do this after I have tested everything.

**Pull request closed without merging**